SI-9314 Marginal edge case to warn-missing-interp

som-snytt opened the pull request:
Don't warn on `"$package_name"` or `"${}"`.  Detecting empty expression also nullifies the test. As in "jury nullification."

som-snytt committed:
SI-9314 Don't warn on "$pkg"

Edge cases of things not to warn about
include package names.

som-snytt committed:
SI-9314 Ignore "${}"

As an Easter egg, let "${} $x" forego the check on `x`.

In other words, empty expression interpolation looks too
degenerate to check.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -105,7 +105,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
   private final val SYNTHETIC_PRIVATE = TRANS_FLAG
 
-  private final val InterpolatorCodeRegex  = """\$\{.*?\}""".r
+  private final val InterpolatorCodeRegex  = """\$\{(.*?)\}""".r
```
Could we simplify `.*?` to `.*`

i won't add a ? at the end of the above question to avoid confusion..  :)

som-snytt replied:
```
scala> val r = """\$\{(.*?)\}""".r
r: scala.util.matching.Regex = \$\{(.*?)\}

scala> r findFirstIn "x ${a} y ${b}"
res0: Option[String] = Some(${a})

scala> val r = """\$\{(.*)\}""".r
r: scala.util.matching.Regex = \$\{(.*)\}

scala> r findFirstIn "x ${a} y ${b}"
res1: Option[String] = Some(${a} y ${b})
```

Ideally, we'd parse and then detect if it parses and any idents are in context.

lrytz replied:
got it thanks.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5206,16 +5206,18 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-            warn("detected an interpolated expression") // "${...}"
+            suspiciousExpr filter (!_.group(1).trim.isEmpty) foreach (_ =>
+              warn("detected an interpolated expression") // "${...}"
+            )
```
I thought you were going to apply the `suspiciousSym` test if the text is an identifier. Then `${junk}` gets a pass like `$pass`. That might solve most conflicts with templating.

som-snytt committed:
SI-9314 No warn on ${nonid}

Use the sym test on an expr that happens to be a subset of
idents and is not in scope. Other `${ operator_* }` warn.

som-snytt commented:
Added not warning on `"${ RANDOM }"` where the simple ident is not in scope. Maybe this covers the templating use case. It takes the edge off the edge case.

lrytz commented:
LGTM, thank you!
